### PR TITLE
Added spec for the feed endpoint

### PR DESCRIPTION
**Pull request by Pchelolo**
We've been talking about adding a schema to the feed endpoint so that some tests could validate agains it - so here's the first iteration of the schema. Depends on https://github.com/wikimedia/restbase/pull/668, after that's merged I will create a test that would validate responses agains this schema and make some minor improvements to the schema itself.

cc @wikimedia/services @wikimedia/mobile 
Bug: https://phabricator.wikimedia.org/T141943

### CODE REVIEW

File: v1/feed.yaml [hunk 1]
```diff
@@ -97,3 +97,141 @@ paths:
...
+      normalizedtitle:
+        type: string
+        description: Article title as it should be presented to the user
+      views:
+        type: integer
+        description: Number of views on the requested day
+      rank:
+        type: integer
+        description: Position in the list of most viewed articles
+      fromencoded:
```
Comment (Pchelolo) [inline v1/feed.yaml]: Q for MCS - what's that?

File: v1/feed.yaml [hunk 2]
```diff
@@ -97,3 +97,182 @@ paths:
...
+    required:
+      - title
+
+  mostread_article:
+    type: object
+    properties:
+      title:
+        type: string
+        description: Article title in a form of DB key
+      normalizedtitle:
```
Comment (gwicke) [inline v1/feed.yaml]: Is this already in the current output / expected by the app?

Comment (Pchelolo) [inline v1/feed.yaml]: This is in the current output, but I don't know how the app uses these

File: test/features/schema_tests.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// mocha defines to avoid JSHint breakage
+/* global describe, it, before, beforeEach, after, afterEach */
+
+var assert = require('../utils/assert.js');
+var server = require('../utils/server.js');
+var preq   = require('preq');
+var Ajv = require('ajv');
+
+describe('Responces should conform to the provided JSON schema of the responce', function() {
```
Comment (berndsi) [inline test/features/schema_tests.js]: typo in response

File: v1/feed.yaml [hunk 4]
```diff
@@ -97,3 +97,182 @@ paths:
                   width: /.+/
                   height: /.+/
 
+definitions:
+  # A https://tools.ietf.org/html/draft-nottingham-http-problem
+  problem:
```
Comment (berndsi) [inline v1/feed.yaml]: So, the 'problem' definition is for error responses?

File: v1/feed.yaml (hunk 2 above)
Comment (berndsi) [inline v1/feed.yaml]: We use the normalized title in the UI when we display the page title. Users don't want to look at titles with underscores, etc.

File: v1/feed.yaml [hunk 5]
```diff
@@ -97,3 +97,182 @@ paths:
...
+        description: Lead image for the article
+        $ref: '#/definitions/thumbnail'
+      description:
+        type: string
+        description: Wikidata description of the article
+      extract:
+        type: string
+        description: First several sentences of an article in plain text
+    required:
+      - title
```
Comment (berndsi) [inline v1/feed.yaml]: Actually, normalizedtitle is also required.

File: v1/feed.yaml (hunk 1 above)
Comment (berndsi) [inline v1/feed.yaml]: Good question. I had not seen this until now. I cannot find the string `fromencoded` in our code base. I guess this is coming in from the PageView API. cc: @mdholloway

File: v1/feed.yaml [hunk 6]
```diff
@@ -97,3 +97,182 @@ paths:
...
+    type: array
+    items:
+      $ref: '#/definitions/news_item'
+
+  random:
+    type: object
+    properties:
+      items:
+        type: array
+        items:
```
Comment (berndsi) [inline v1/feed.yaml]: The two levels of `items` looks suspicious.

File: v1/feed.yaml (hunk 4 above)
Comment (Pchelolo) [inline v1/feed.yaml]: Yep, it's the same everywhere in the API

File: v1/feed.yaml (hunk 6 above)
Comment (Pchelolo) [inline v1/feed.yaml]: The first `items` is the property name, second `items` is a keyword in JSON schema.

Right now random portion is returned in the form of
```
random:
  items:
    - title: 'ads'
      
```

why is that? Why don't we change it to
```
random:
  title: 'sdcasdc'
```
?

File: v1/feed.yaml (hunk 1 above)
Comment (berndsi) [inline v1/feed.yaml]: @mdholloway is not around today. Maybe @milimetric recognizes the `fromencoded` property and what it represents? I'm leaning towards leaving this out from the spec for now unless one of them says that it's important.

File: v1/feed.yaml (hunk 6 above)
Comment (berndsi) [inline v1/feed.yaml]: I was leaning towards skipping the items array as well. It was just introduced to match the RB format of the /page/title/{title} endpoint response. The one point that puts me back on the fence about this would be that maybe in the future we might want to return multiple random articles.

Comment (Pchelolo) [inline v1/feed.yaml]: @berndsi good point about multiple random articles...

**Comment by Pchelolo** [comment]
Updated according to comments, rebased.

File: v1/feed.yaml (hunk 6 above)
Comment (gwicke) [inline v1/feed.yaml]: We could just use a different key for different kinds of random titles. My personal preference would be to keep things flat, if that's possible without breaking existing clients.

The main reason why RB is using `items` in listing responses is that this sets up a top-level object, which allows adding continuations & other attributes later.

Comment (Pchelolo) [inline v1/feed.yaml]: @gwicke using items here would be in line with other RB endpoints and would allow adding more random articles without paint. 

If you use different keys, how would you add 5 random articles instead of 1? `random`, `random2`, `random3` ... ? Kinda ugly. Since it doesn't really matter let's leave it as is?

File: v1/feed.yaml (hunk 5 above)
Comment (Pchelolo) [inline v1/feed.yaml]: done

File: v1/feed.yaml (hunk 6 above)
Comment (Pchelolo) [inline v1/feed.yaml]: @gwicke Ok, converted `random` portion to an array

**Comment by gwicke** [comment]
LGTM.

File: v1/feed.js [hunk 7]
```diff
@@ -193,7 +197,7 @@ Feed.prototype.aggregated = function(hyper, req) {
             }
 
             if (feed.random) {
-                feed.random.items = assignAllSummaries(feed.random.items);
+                feed.random = assignAllSummaries(feed.random.items);
```
Comment (d00rman) [inline v1/feed.js]: Uh, the Android app expects to have `feed.random.items`. Why this change in the interface?

Comment (gwicke) [inline v1/feed.js]: I asked the same question, but it seems that the Android app is not using this property yet, and IOS will be the first to start using this soon. Would be good to have confirmation from @berndsi, though.

Comment (berndsi) [inline v1/feed.js]: That is correct. The Android app is not using this property yet, and IOS will be the first to start using this.

File: v1/feed.yaml [hunk 8]
```diff
@@ -97,3 +96,178 @@ paths:
...
+        description: Data about the featured article of the day
+        $ref: '#/definitions/article_summary'
+      mostread:
+        description: Data about most viewed articles
+        $ref: '#/definitions/mostread'
+      news:
+        description: Data about in the news section
+        $ref: '#/definitions/news'
+      random:
+        description: A single random article
```
Comment (d00rman) [inline v1/feed.yaml]: Aren't there multiple ones?

Comment (Pchelolo) [inline v1/feed.yaml]: @d00rman Right now there's only one
